### PR TITLE
vendor._altera: implement `lib.io` buffer primitives.

### PR DESCRIPTION
**Pull request by wanda-phi**

### CODE REVIEW

**Review by whitequark — APPROVED** [review]

**Comment by tpwrules** [comment]
Switched my design to this branch and it works well with Quartus 20.1.1.720. It only uses basic I/O though, not even bidirectional.

File: amaranth/vendor/_altera.py [hunk 1]
```diff
@@ -293,247 +445,21 @@ def create_missing_domain(self, name):
...
-        return m
+    def get_io_buffer(self, buffer):
+        if isinstance(buffer, io.Buffer):
+            res = IOBuffer(buffer.direction, buffer.port)
+        elif isinstance(buffer, io.FFBuffer):
+            res = FFBuffer(buffer.direction, buffer.port)
+        elif isinstance(buffer, io.DDRBuffer):
+            res = DDRBuffer(buffer.direction, buffer.port)
+        else:
+            raise TypeError(f"Unsupported buffer type {buffer!r}")
```
Comment (whitequark) [inline amaranth/vendor/_altera.py:456]:
```suggestion
            raise TypeError(f"Unsupported buffer type {buffer!r}") # :nocov:
```

File: amaranth/vendor/_altera.py [hunk 2]
```diff
@@ -1,9 +1,161 @@
...
+            if direction is not io.Direction.Input:
+                self.oe.attrs["useioff"] = 1
+
+    def elaborate(self, platform):
+        kwargs = dict(
+            p_enable_bus_hold="FALSE",
+            p_number_of_channels=len(self.port),
+        )
+        if isinstance(self.port, io.SingleEndedPort):
+            kwargs["p_use_differential_mode"] = "FALSE"
```
Comment (whitequark) [inline amaranth/vendor/_altera.py:46]: You don't actually need to use this--[`Instance` accepts arguments of the form `("p", "blah", value)`](https://amaranth-lang.org/docs/amaranth/latest/guide.html#instances).

**Review by whitequark — APPROVED** [review]
LGTM. This is still not perfect but the Altera uses will simply share our suffering.

**Pull request merged**